lib/elixir/lib/inspect.ex

```diff
@@ -478,6 +481,15 @@ defimpl Inspect, for: Any do
     end
   end
 
+  defp validate_option(option_list, fields, module, caller) do
+    if not Enum.empty?(option_list -- fields) do
+      description = "When deriving inspect protocol of #{module}, values must match struct fields"
+      raise CompileError, file: caller.file, line: caller.line, description: description
```

Comment:
This can be a regular ArgumentError :) CompileError is typically only raise from within the compiler.

Reply:
It would also be nice to show which fields do not match. Perhaps convert the `if` to a case and pattern match accordingly.

Reply:
updated dc203bfe0c8d9f9999f59b0897056f4e2ccccb2a